{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9b4b5f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This script rename and reorganize the original fastq files downloaded from the Genome Sequence Archive \n",
    "https://ngdc.cncb.ac.cn, accession number HRA001730. Make these files ready for cellranger count alignment\n",
    "\n",
    "author: Zhiyuan Yao\n",
    "\n",
    "Update Date: Aug-10-2022\n",
    "\"\"\"\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from copy import copy\n",
    "import os\n",
    "import glob\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1952d481",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filename_reorg(raw_info, raw_info_2):\n",
    "    \"\"\"Rename fastq files to meet cellranger's standard\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    raw_info: dataframe\n",
    "        The pandas dataframe that stores the fastq original file name, Accession number, corresponding experimental\n",
    "        accession number and Run title etc. \n",
    "    raw_info_2: dataframe\n",
    "        The pandas dataframe that stores experimental accession number and its corresponding Sample name and Run title\n",
    "        raw_info and raw_info_2 are extracted from the excel table file downloaded from the Genome Sequence Archive \n",
    "        https://ngdc.cncb.ac.cn, accession number HRA001730\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    raw_info: dataframe\n",
    "        The pandas dataframe that stores old names and new names for each fastq files\n",
    "    \"\"\"\n",
    "    #get the lane numbers\n",
    "    lane_number = raw_info['Run title'].str[-1]\n",
    "    \n",
    "    #add one column of sample names\n",
    "    for ind in raw_info.index:\n",
    "        exp_id = raw_info.loc[ind, 'Experiment accession']\n",
    "        raw_info.loc[ind, 'sample_name'] = raw_info_2.loc[exp_id, 'BioSample name']\n",
    "        \n",
    "    \n",
    "    raw_info['new_name_R1'] = raw_info['sample_name']+'_S1_L00'+lane_number+'_R1_001.fastq.gz'\n",
    "    raw_info['new_name_R2'] = raw_info['sample_name']+'_S1_L00'+lane_number+'_R2_001.fastq.gz'\n",
    "    \n",
    "    return(raw_info)\n",
    "    \n",
    "def make_input_folder(raw_info,path):\n",
    "    \"\"\"Generate input folders to store renamed fastq files\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    raw_info: dataframe\n",
    "        The pandas dataframe that stores old names and new names for each fastq files, generated from the function \n",
    "        filename_reorg  \n",
    "    path: str\n",
    "        The root path where the input folders to be created\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"    \n",
    "    print(\"making input folders...\")\n",
    "    sample_list = list(raw_info['sample_name'].unique())\n",
    "    for sample in sample_list:\n",
    "        folder_dir = os.path.join(path, sample)\n",
    "        if not os.path.isdir(folder_dir):\n",
    "            os.mkdir(folder_dir)\n",
    "        else:\n",
    "            raise Exception(\"The folder already exists\")\n",
    "    print(\"{} input folders has been created\".format(len(sample_list)))\n",
    "        \n",
    "def mv_input_file(raw_info, src_path, dst_path):\n",
    "    \"\"\"Rename and move the fastq files to the corresponding input folders\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    raw_info: dataframe\n",
    "        The pandas dataframe that stores old names and new names for each fastq files, generated from the function \n",
    "        filename_reorg  \n",
    "    src_path: str\n",
    "        The path where the original fastq files located\n",
    "    dst_path: str\n",
    "        The root path where the input folders to be created\n",
    "        \n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\" \n",
    "    print(\"moving files from {} to {}\".format(src_path, dst_path))\n",
    "    for ind in raw_info.index:\n",
    "        filename_1 = raw_info.loc[ind, 'File name 1']\n",
    "        new_filename_1 = raw_info.loc[ind, 'new_name_R1']\n",
    "        filename_2 = raw_info.loc[ind, 'File name 2']\n",
    "        new_filename_2 = raw_info.loc[ind, 'new_name_R2']\n",
    "        \n",
    "        s_id = raw_info.loc[ind, 'sample_name']\n",
    "        \n",
    "        file_src_path_1 = os.path.join(src_path, filename_1)\n",
    "        file_dst_path_1 = os.path.join(dst_path, s_id, new_filename_1)\n",
    "        \n",
    "        file_src_path_2 = os.path.join(src_path, filename_2)\n",
    "        file_dst_path_2 = os.path.join(dst_path, s_id, new_filename_2)\n",
    "        \n",
    "        shutil.copyfile(file_src_path_1, file_dst_path_1)\n",
    "        print(\"file {} has been sucessfully renamed to {} and moved\".format(filename_1, new_filename_1))\n",
    "        shutil.copyfile(file_src_path_2, file_dst_path_2)\n",
    "        print(\"file {} has been sucessfully renamed to {} and moved\".format(filename_2, new_filename_2))\n",
    "        \n",
    "def batch(input_path, batch_file_path, n_sub):\n",
    "    \"\"\"split samples into batches and record the information to a tsv file for each batch\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    input_path: str\n",
    "        The root path where the input folders to be created \n",
    "    batch_file_path: str\n",
    "        The path where the tsv files will be stored\n",
    "    n_sub: int\n",
    "        The number of samples in each batch\n",
    "        \n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\" \n",
    "    path_list = glob.glob(os.path.join(input_path,'*'))\n",
    "    if len(path_list) == 0:\n",
    "        raise IOError('No input subfolders')\n",
    "    sample_list = [x.split('/')[-1] for x in path_list]\n",
    "    \n",
    "    df_temp = pd.DataFrame(sample_list, columns = ['name'])\n",
    "    n_group = len(df_temp.index) // n_sub + 1\n",
    "    \n",
    "    for g_id in np.arange(n_group):\n",
    "        f_name = 's_batch_'+str(g_id)+'.tsv'\n",
    "        if g_id != (n_group-1):\n",
    "            temp = test.iloc[g_id*n_sub : (g_id+1)*n_sub]\n",
    "        else:\n",
    "            temp = test.iloc[g_id*n_sub : ]\n",
    "        \n",
    "        temp.to_csv(os.path.join(batch_file_path, f_name), sep = '\\t')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "99df5a17",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    \n",
    "    #load infomation metrics\n",
    "    raw_info_path_1 = '/domino/edv/id-td-virology/Zhiyuan/public/Gut/processed_data/raw_info.csv'\n",
    "    raw_info_path_2 = '/domino/edv/id-td-virology/Zhiyuan/public/Gut/processed_data/raw_info_2.csv'\n",
    "    raw_info = pd.read_csv(raw_info_path_1, sep = ',', index_col = 0)\n",
    "    raw_info_2 = pd.read_csv(raw_info_path_2, sep = ',', index_col = 0)\n",
    "        \n",
    "    #add new filenames that fit the cellranger formats\n",
    "    info_test = filename_reorg(raw_info, raw_info_2)\n",
    "    \n",
    "    #make folders for mapping-ready files\n",
    "    dst_path = '/domino/edv/id-td-virology/Public_dataset/2022_Gut/input/'\n",
    "    make_input_folder(info_test,dst_path)\n",
    "    \n",
    "    #path where the source files located\n",
    "    src_path = '/domino/edv/id-td-virology/Public_dataset/2022_Gut/rawdata/raw_fastq/'\n",
    "    #move files to the folders \n",
    "    mv_input_file(info_test, src_path, dst_path)\n",
    "    \n",
    "    #split samples into batches\n",
    "    input_path = '/domino/edv/id-td-virology/Public_dataset/2022_Gut/input/'\n",
    "    batch_file_path = '/domino/edv/id-td-virology/Public_dataset/2022_Gut/batch_file/'\n",
    "    batch(input_path, batch_file_path, n_sub = 8)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ce34787",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
