{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3b849ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This script mapped the fastq files to the genome using cellranger, the script is called by the bash file 'run_count.sh'\n",
    "author: Zhiyuan Yao\n",
    "\n",
    "Update Date: Aug-10-2022\n",
    "\"\"\"\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from copy import copy\n",
    "import os\n",
    "import glob\n",
    "import shutil\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c4f1bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def grab_sample_list(input_path, sample_table = None):\n",
    "    \"\"\"Get the sample list which is ready for mapping, if the sample_table is provided, only mapped the sample on the table,\n",
    "    otherwise scan the input folder and generate the sample list\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    input_path: str\n",
    "        The root path where the input folders to be created \n",
    "    sample_table: dataframe\n",
    "        The table including the specified samples, the tables are from the tsv files generated by batch splitting process. \n",
    "\n",
    "    Returns:\n",
    "        sample_list: list\n",
    "        The sample list to be mapped\n",
    "    \"\"\" \n",
    "    if sample_table is None:\n",
    "        print(\"gathering sample lists from the input folder...\")\n",
    "        path_list = glob.glob(os.path.join(input_path,'*'))\n",
    "        if len(path_list) == 0:\n",
    "            raise IOError('No input subfolders')\n",
    "        sample_list = [x.split('/')[-1] for x in path_list]\n",
    "            \n",
    "    else:\n",
    "        print(\"gathering sample lists from the cumstom sample table...\")\n",
    "        sample_list = list(sample_table.name)\n",
    "        \n",
    "    print(\"{} samples are found:\".format(len(sample_list)))\n",
    "    for sample in sample_list:\n",
    "        print(sample)\n",
    "        \n",
    "        \n",
    "    return(sample_list)\n",
    "\n",
    "def scan_input_folder(sample_list, input_path):\n",
    "    \"\"\"Scan the fastq files from the specified samples, check whether they have correct number of R1 and R2 files\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    sample_list: list\n",
    "        The sample list to be mapped\n",
    " \n",
    "    input_path: str\n",
    "        The root path where the input folders to be created \n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    for sample in sample_list:\n",
    "        fn_r1 = glob.glob(os.path.join(input_path,sample,'*R1_001.fastq.gz'))\n",
    "        fn_r2 = glob.glob(os.path.join(input_path,sample,'*R2_001.fastq.gz'))\n",
    "        \n",
    "        if (len(fn_r1)!=len(fn_r2)):\n",
    "            raise IOError('Sample {} does not have equal R1 and R2 fastqs'.format(sample))\n",
    "        elif ((len(fn_r1) == 0) & (len(fn_r2) == 0)):\n",
    "            raise IOError('Sample {} has no correctly re-named fastqs'.format(sample))\n",
    "        else:\n",
    "            print('Sample {} has correct fastqs'.format(sample))\n",
    "            \n",
    "    print('scan input folder completed, {} samples are ready to map'.format(len(sample_list)))\n",
    "            \n",
    "            \n",
    "\n",
    "def make_output_folder(output_path, sample_list):\n",
    "    \"\"\"Generate output folders for the mapping results\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    output_path: str\n",
    "        The root path where the ouput folders to be created     \n",
    "    sample_list: list\n",
    "        The sample list to be mapped\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    print(\"make output folders...\")\n",
    "    for sample in sample_list:\n",
    "        print(\"creating the folder for sample {}\".format(sample))\n",
    "        folder_dir = os.path.join(output_path, sample)\n",
    "        if not os.path.isdir(folder_dir):\n",
    "            os.mkdir(folder_dir)\n",
    "        else:\n",
    "            raise Exception(\"The folder already exists\")\n",
    "    print(\"{} output folders has been created\".format(len(sample_list)))\n",
    "    \n",
    "def cellranger_count(sample_list, input_path, output_path, cores = 40, mem = 64):  \n",
    "    \"\"\"Run cell ranger count\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    sample_list: list\n",
    "        The sample list to be mapped    \n",
    "    input_path: str\n",
    "        The root path where the input folders to be created         \n",
    "    output_path: str\n",
    "        The root path where the ouput folders to be created \n",
    "    cores: int\n",
    "        max numbers of cpus to be used\n",
    "    mem: int\n",
    "        max GB memory to be used\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    os.chdir(output_path)\n",
    "    c = 0\n",
    "    f = []\n",
    "    for sample in sample_list:\n",
    "        print(\"start mapping {}\".format(sample))\n",
    "        fastqs_path = os.path.join(input_path, sample)\n",
    "        \n",
    "#         cellranger_command = [\"cellranger\",\n",
    "#                               \"count\",\n",
    "#                               \"--id={}\".format(sample),\n",
    "#                               \"--transcriptome=/domino/edv/id-td-virology/Aligner/cellranger/refdata-gex-GRCh38-2020-A\",\n",
    "#                               \"--fastqs={}\".format(fastqs_path),\n",
    "#                               \"--sample={}\".format(sample),\n",
    "#                               \"--include-introns=false\",\n",
    "#                               \"--localcores={}\".format(cores),\n",
    "#                               \"--localmem={}\".format(mem)]\n",
    "        \n",
    "        cellranger_command = \"cellranger count \\\n",
    "--id={} \\\n",
    "--transcriptome=/domino/edv/id-td-virology/Aligner/cellranger/refdata-gex-GRCh38-2020-A \\\n",
    "--fastqs={} \\\n",
    "--sample={} \\\n",
    "--include-introns=false \\\n",
    "--localcores={} \\\n",
    "--localmem={}\".format(sample, fastqs_path, sample, cores, mem)\n",
    "                \n",
    "        a = subprocess.run(cellranger_command, capture_output=True, check = False, text = True, shell = True)        \n",
    "        if a.returncode == 0:\n",
    "            print('sample {} sucessfully processed'.format(sample))\n",
    "            c = c + 1\n",
    "        else:\n",
    "            print(\"sample {} processing failed\".format(sample))\n",
    "            f.append(sample)\n",
    "            continue\n",
    "    \n",
    "    print(\"mapping process is completed\")        \n",
    "    print(\"{} samples are sucessfully processed, {} samples failed\".format(c, len(f)))\n",
    "    if len(f) != 0:\n",
    "        print(\"The failed samples:\")\n",
    "        for f_sample in f:\n",
    "            print(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec048474",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    #root path of the input folders\n",
    "    input_path = '/domino/edv/id-td-virology/Public_dataset/2022_Gut/input/'\n",
    "    #root path of the output folders\n",
    "    output_path = '/domino/edv/id-td-virology/Public_dataset/2022_Gut/output/'\n",
    "    #files contain sample information from one batch, can be skipped if sample_table == none\n",
    "    batch_path = '/domino/edv/id-td-virology/Public_dataset/2022_Gut/batch_file/s_batch_5.tsv'\n",
    "    sample_table = pd.read_csv(batch_path, sep = '\\t', index_col = 0)\n",
    "    \n",
    "    #generate sample list\n",
    "    sample_list = grab_sample_list(input_path, sample_table = sample_table)\n",
    "    #check the fastq files in the input folder are in the correct numbers\n",
    "    scan_input_folder(sample_list, input_path)\n",
    "    #generate the output folders\n",
    "    make_output_folder(output_path, sample_list)\n",
    "    #run cellranger count\n",
    "    cellranger_count(sample_list, input_path, output_path, cores = 24, mem = 128)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3056a45",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
